ir/src/symbol_table/symbol.rs

```diff
@@ -49,121 +49,189 @@ impl Symbol {
-        access_type: AccessType,
+        symbol_access: SymbolAccess,
     ) -> Result<Value, SemanticError> {
-        match access_type {
+        if symbol_access.offset() != 0 {
+            return Err(SemanticError::invalid_access_offset(
+                self,
+                symbol_access.offset(),
+            ));
+        }
+
+        match symbol_access.access_type() {
             AccessType::Default => {
                 if binding_size != 1 {
-                    return Err(SemanticError::invalid_random_value_binding_access(
-                        self.name(),
+                    return Err(SemanticError::invalid_access_type(
+                        self,
+                        symbol_access.access_type(),
                     ));
                 }
                 Ok(Value::RandomValue(binding_offset))
             }
             AccessType::Vector(idx) => {
-                if idx >= binding_size {
-                    return Err(SemanticError::vector_access_out_of_bounds(
-                        self.name(),
-                        idx,
-                        binding_size,
+                if *idx >= binding_size {
+                    return Err(SemanticError::invalid_access_type(
+                        self,
+                        symbol_access.access_type(),
                     ));
                 }
 
                 let offset = binding_offset + idx;
                 Ok(Value::RandomValue(offset))
             }
-            _ => Err(SemanticError::invalid_random_value_access_type(
-                self.name(),
-                &access_type,
+            _ => Err(SemanticError::invalid_access_type(
+                self,
+                symbol_access.access_type(),
             )),
         }
     }
 
     fn get_trace_value(
         &self,
         binding: &TraceBinding,
-        access_type: AccessType,
+        symbol_access: SymbolAccess,
     ) -> Result<Value, SemanticError> {
-        // symbol accesses at rows other than the first are identified by the parser as
-        // [NamedTraceAccess] and handled differently, so this case will only occur for
-        // trace column accesses at the current row.
-        // TODO: can we handle this differently so it's more explicit & get rid of this comment?
-        let row_offset = CURRENT_ROW;
+        let (_, access_type, offset) = symbol_access.into_parts();
         match access_type {
             AccessType::Default => {
                 if binding.size() != 1 {
-                    return Err(SemanticError::invalid_trace_binding_access(self.name()));
+                    return Err(SemanticError::invalid_access_type(self, &access_type));
                 }
                 let trace_segment = binding.trace_segment();
                 let trace_access =
-                    TraceAccess::new(trace_segment, binding.offset(), binding.size(), row_offset);
+                    TraceAccess::new(trace_segment, binding.offset(), binding.size(), offset);
```

Comment:
Two different offsets are a bit confusing here. Maybe renaming `offset` variable here into `row_offset` would help a bit? Though it still may not be clear that `binding.offset()` is a column offset.